protostar/starknet/cheaters/contracts.py

```diff
@@ -311,3 +308,49 @@ async def invoke(
                 state=state_copy,
                 general_config=StarknetGeneralConfig(),
             )
+
+    async def send_message_to_l2(
+        self,
+        selector: Selector,
+        from_l1_address: Address,
+        to_l2_address: Address,
+        payload: Optional[CairoOrPythonData] = None,
+    ) -> None:
+        cairo_calldata = await self._transform_payload_from_l1_to_cairo_data_by_addr(
+            l1_address=from_l1_address,
+            contract_address=to_l2_address,
+            payload=payload,
+            selector=selector,
+        )
+        entry_point = ExecuteEntryPoint.create_for_testing(
+            contract_address=int(to_l2_address),
+            calldata=cairo_calldata,
+            caller_address=int(from_l1_address),
+            entry_point_selector=int(selector),
+            entry_point_type=EntryPointType.L1_HANDLER,
+            call_type=CallType.DELEGATE,
+            class_hash=await self.cheatable_state.get_class_hash_at(int(to_l2_address)),
+        )
+        with self.cheatable_state.copy_and_apply() as state_copy:
+            await entry_point.execute_for_testing(
+                state=state_copy,
+                general_config=StarknetGeneralConfig(),
+            )
+
+    async def _transform_payload_from_l1_to_cairo_data_by_addr(
+        self,
+        contract_address: Address,
+        l1_address: Address,
+        selector: Selector,
+        payload: Optional[CairoOrPythonData] = None,
+    ) -> CairoData:
+        if payload is None:
+            return []
+        if isinstance(payload, collections.Mapping):
+            python_payload = {"from_address": int(l1_address), **payload}
+            return await self._transform_calldata_to_cairo_data_by_addr(
+                function_name=str(selector),
+                calldata=python_payload,
+                contract_address=contract_address,
```

Comment:
Perhaps we should consider transferring this functionality to another cheater, along with call, invoke - a transaction-based cheater. I'd do this in another PR though, after #1368

Reply:
Maybe in a different PR.

Reply:
Maybe name it L1Cheater?
I agree that Contracts cheater is not a good location

Reply:
Can be done in a separate PR, but please create a task for it

Reply:
I'd have to move the data transformation code out of this class, and that's a subject for a separate PR.

Reply:
@karol-bisztyga already did it in #1363 , you can repoint the PR into it

Reply:
Removed support for data transformation. I don't know if there will be a resonable way to use it in Cairo 1. If it's needed, it can be added in a different PR because this one hangs already too long.